File: resources/xcode/NativePHP/Queue/NativeQueueCoordinator.swift
import Foundation

/// Coordinates background queue processing for Laravel jobs.
///
/// The coordinator monitors the job queue and triggers processing during idle periods.
/// It ensures jobs don't interfere with UI responsiveness by:
/// - Processing one job at a time
/// - Pausing during active user interaction
/// - Throttling based on configuration
///
/// Jobs are processed via HTTP requests to `/_native/queue/work`, which executes
/// one job and returns. The coordinator decides when to process the next job.
final class NativeQueueCoordinator {
    static let shared = NativeQueueCoordinator()

    // MARK: - Configuration

    /// Minimum delay between job processing attempts (seconds)
    var minDelayBetweenJobs: TimeInterval = 0.1

    /// Delay when queue is empty before checking again (seconds)
    var emptyQueuePollInterval: TimeInterval = 2.0

    /// Maximum jobs to process per batch before yielding (0 = unlimited)
    var maxJobsPerBatch: Int = 10

    /// Whether to pause processing during active scrolling
    var pauseDuringScroll: Bool = true

    // MARK: - State

    private var isRunning = false
    private var isPaused = false
    private var pendingJobCount = 0
    private var processedInBatch = 0
    private var totalProcessed = 0

    private let processingQueue = DispatchQueue(
        label: "com.nativephp.queue.coordinator",
        qos: .utility
    )

    private var workTimer: DispatchSourceTimer?

    private init() {}

    // MARK: - Public API

    /// Start the queue coordinator.
    /// Called automatically when the app launches.
    func start() {
        processingQueue.async { [weak self] in
            guard let self = self, !self.isRunning else { return }

            self.isRunning = true
            self.processedInBatch = 0
            self.checkForJobs()
        }
    }

    /// Stop the queue coordinator.
    func stop() {
        processingQueue.async { [weak self] in
            guard let self = self else { return }

            self.isRunning = false
            self.workTimer?.cancel()
            self.workTimer = nil
        }
    }

    /// Pause job processing (e.g., during heavy UI activity).
    func pause() {
        processingQueue.async { [weak self] in
            self?.isPaused = true
        }
    }

    /// Resume job processing.
    func resume() {
        processingQueue.async { [weak self] in
            guard let self = self else { return }

            self.isPaused = false

            if self.isRunning && self.pendingJobCount > 0 {
                self.scheduleNextJob(delay: self.minDelayBetweenJobs)
            }
        }
    }

    /// Notify that jobs are available (called from Queue.JobsAvailable bridge function).
    func notifyJobsAvailable(count: Int, queue: String) {
        processingQueue.async { [weak self] in
            guard let self = self else { return }

            let wasEmpty = self.pendingJobCount == 0
            self.pendingJobCount = count

            // If we were idle and now have jobs, start processing
            if wasEmpty && count > 0 && self.isRunning && !self.isPaused {
                self.scheduleNextJob(delay: self.minDelayBetweenJobs)
            }
        }
    }

    // MARK: - Private

    private func checkForJobs() {
        guard isRunning else { return }

        // Make status request to get current job count and config
        makeRequest(path: "/_native/queue/status") { [weak self] result in
            guard let self = self else { return }

            switch result {
            case .success(let data):
                if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let pending = json["pending"] as? Int {
                    self.pendingJobCount = pending

                    // Apply config from PHP if available
                    if let config = json["config"] as? [String: Any] {
                        if let minDelay = config["min_delay"] as? Int {
                            self.minDelayBetweenJobs = TimeInterval(minDelay) / 1000.0
                        }
                        if let batchSize = config["batch_size"] as? Int {
                            self.maxJobsPerBatch = batchSize
                        }
                        if let pollInterval = config["poll_interval"] as? Int {
                            self.emptyQueuePollInterval = TimeInterval(pollInterval) / 1000.0
                        }
                    }

                    if pending > 0 && !self.isPaused {
                        self.scheduleNextJob(delay: self.minDelayBetweenJobs)
                    } else {
                        // Queue is empty, poll again later
                        self.scheduleNextJob(delay: self.emptyQueuePollInterval)
                    }
                }

            case .failure:
                // Retry after a delay
                self.scheduleNextJob(delay: self.emptyQueuePollInterval)
            }
        }
    }

    private func scheduleNextJob(delay: TimeInterval) {
        workTimer?.cancel()

        let timer = DispatchSource.makeTimerSource(queue: processingQueue)
        timer.schedule(deadline: .now() + delay)
        timer.setEventHandler { [weak self] in
            self?.processNextJob()
        }
        timer.resume()
        workTimer = timer
    }

    private func processNextJob() {
        guard isRunning, !isPaused else { return }

        // Check batch limit
        if maxJobsPerBatch > 0 && processedInBatch >= maxJobsPerBatch {
            processedInBatch = 0
            scheduleNextJob(delay: emptyQueuePollInterval)
            return
        }

        // Process one job
        makeRequest(path: "/_native/queue/work", method: "POST") { [weak self] result in
            guard let self = self, self.isRunning else { return }

            switch result {
            case .success(let data):
                self.handleWorkResponse(data)

            case .failure:
                self.scheduleNextJob(delay: self.emptyQueuePollInterval)
            }
        }
    }

    private func handleWorkResponse(_ data: Data) {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            scheduleNextJob(delay: emptyQueuePollInterval)
            return
        }

        let processed = json["processed"] as? Bool ?? false
        let pending = json["pending"] as? Int ?? 0

        self.pendingJobCount = pending

        if processed {
            processedInBatch += 1
            totalProcessed += 1

            // Extract job info
            let jobData = json["job"] as? [String: Any]
            let jobId = jobData?["id"] as? String ?? "unknown"
            let jobName = jobData?["name"] as? String ?? "unknown"
            let queue = jobData?["queue"] as? String ?? "default"
            let durationMs = json["duration_ms"] as? Int ?? 0

            // Check for error (job failed but was processed)
            if let error = json["error"] as? [String: Any] {
                let errorMessage = error["message"] as? String ?? "Unknown error"

                // Dispatch failure event
                LaravelBridge.shared.send?("Native\\Mobile\\Queue\\Events\\JobFailed", [
                    "jobId": jobId,
                    "jobName": jobName,
                    "error": errorMessage,
                    "willRetry": true,
                    "attempts": jobData?["attempts"] as? Int ?? 1,
                    "pending": pending
                ])
            } else {
                // Dispatch success event
                LaravelBridge.shared.send?("Native\\Mobile\\Queue\\Events\\JobCompleted", [
                    "jobId": jobId,
                    "jobName": jobName,
                    "queue": queue,
                    "durationMs": durationMs,
                    "pending": pending
                ])
            }

            // More jobs? Process next one
            if pending > 0 {
                scheduleNextJob(delay: minDelayBetweenJobs)
            } else {
                // Dispatch queue empty event
                LaravelBridge.shared.send?("Native\\Mobile\\Queue\\Events\\QueueEmpty", [
                    "queue": jobData?["queue"] as? String ?? "default",
                    "processedCount": totalProcessed
                ])

                // Poll again later in case new jobs arrive
                scheduleNextJob(delay: emptyQueuePollInterval)
            }
        } else {
            // Queue was empty
            scheduleNextJob(delay: emptyQueuePollInterval)
        }
    }

    // MARK: - HTTP Helpers

    private func makeRequest(
        path: String,
        method: String = "GET",
        completion: @escaping (Result<Data, Error>) -> Void
    ) {
        // Execute through PHP on the shared serial queue to prevent concurrent PHP execution
        // Uses the same queue as PHPSchemeHandler to ensure WebView and queue requests don't overlap
        PHPSchemeHandler.phpSerialQueue.async {
            let requestData = RequestData(
                method: method,
                uri: path,
                data: nil,
                query: nil,
                headers: [
                    "Accept": "application/json",
                    "Content-Type": "application/json"
                ]
            )

            guard let response = NativePHPApp.laravel(request: requestData) else {
                completion(.failure(NSError(domain: "NativeQueueCoordinator", code: -2,
                                           userInfo: [NSLocalizedDescriptionKey: "No response from PHP"])))
                return
            }

            // Parse response - extract body from HTTP response
            let components = response.components(separatedBy: "\r\n\r\n")
            if components.count >= 2 {
                let body = components[1]
                if let data = body.data(using: .utf8) {
                    completion(.success(data))
                    return
                }
            }

            completion(.failure(NSError(domain: "NativeQueueCoordinator", code: -3,
                                       userInfo: [NSLocalizedDescriptionKey: "Invalid response format"])))
        }
    }
}

File: resources/androidstudio/app/src/main/java/com/nativephp/mobile/queue/NativeQueueCoordinator.kt
package com.nativephp.mobile.queue

import android.os.Handler
import android.os.Looper
import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.PHPBridge
import com.nativephp.mobile.network.PHPRequest
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONObject
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Coordinates background queue processing for Laravel jobs.
 *
 * The coordinator monitors the job queue and triggers processing during idle periods.
 * It ensures jobs don't interfere with UI responsiveness by:
 * - Processing one job at a time
 * - Pausing during active user interaction
 * - Throttling based on configuration
 *
 * Jobs are processed via HTTP requests to `/_native/queue/work`, which executes
 * one job and returns. The coordinator decides when to process the next job.
 */
class NativeQueueCoordinator private constructor() {

    companion object {
        @Volatile
        private var instance: NativeQueueCoordinator? = null

        fun getInstance(): NativeQueueCoordinator {
            return instance ?: synchronized(this) {
                instance ?: NativeQueueCoordinator().also { instance = it }
            }
        }
    }

    // Configuration
    /** Minimum delay between job processing attempts (milliseconds) */
    var minDelayBetweenJobs: Long = 100

    /** Delay when queue is empty before checking again (milliseconds) */
    var emptyQueuePollInterval: Long = 2000

    /** Maximum jobs to process per batch before yielding (0 = unlimited) */
    var maxJobsPerBatch: Int = 10

    /** Whether to pause processing during active scrolling */
    var pauseDuringScroll: Boolean = true

    // Dependencies
    private var phpBridge: PHPBridge? = null
    private var activityProvider: (() -> FragmentActivity?)? = null

    // State
    private val isRunning = AtomicBoolean(false)
    private val isPaused = AtomicBoolean(false)
    private val pendingJobCount = AtomicInteger(0)
    private var processedInBatch = 0
    private var totalProcessed = 0

    private val executor = Executors.newSingleThreadScheduledExecutor()
    private var scheduledTask: ScheduledFuture<*>? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Initialize the coordinator with required dependencies.
     * Call this from MainActivity.onCreate().
     */
    fun initialize(phpBridge: PHPBridge, activityProvider: () -> FragmentActivity?) {
        this.phpBridge = phpBridge
        this.activityProvider = activityProvider
    }

    /**
     * Start the queue coordinator.
     * Called automatically when the app launches.
     */
    fun start() {
        if (isRunning.getAndSet(true)) {
            return // Already running
        }

        processedInBatch = 0
        checkForJobs()
    }

    /**
     * Stop the queue coordinator.
     */
    fun stop() {
        isRunning.set(false)
        scheduledTask?.cancel(false)
        scheduledTask = null
    }

    /**
     * Pause job processing (e.g., during heavy UI activity).
     */
    fun pause() {
        isPaused.set(true)
    }

    /**
     * Resume job processing.
     */
    fun resume() {
        isPaused.set(false)

        if (isRunning.get() && pendingJobCount.get() > 0) {
            scheduleNextJob(minDelayBetweenJobs)
        }
    }

    /**
     * Notify that jobs are available (called from Queue.JobsAvailable bridge function).
     */
    fun notifyJobsAvailable(count: Int, queue: String) {
        val wasEmpty = pendingJobCount.get() == 0
        pendingJobCount.set(count)

        // If we were idle and now have jobs, start processing
        if (wasEmpty && count > 0 && isRunning.get() && !isPaused.get()) {
            scheduleNextJob(minDelayBetweenJobs)
        }
    }

    // MARK: - Private

    private fun checkForJobs() {
        if (!isRunning.get()) return

        // Make status request to get current job count and config
        makeRequest("/_native/queue/status", "GET") { result ->
            result.fold(
                onSuccess = { json ->
                    val pending = json.optInt("pending", 0)
                    pendingJobCount.set(pending)

                    // Apply config from PHP if available
                    val config = json.optJSONObject("config")
                    if (config != null) {
                        minDelayBetweenJobs = config.optLong("min_delay", 100)
                        maxJobsPerBatch = config.optInt("batch_size", 10)
                        emptyQueuePollInterval = config.optLong("poll_interval", 2000)
                    }

                    if (pending > 0 && !isPaused.get()) {
                        scheduleNextJob(minDelayBetweenJobs)
                    } else {
                        // Queue is empty, poll again later
                        scheduleNextJob(emptyQueuePollInterval)
                    }
                },
                onFailure = {
                    // Retry after a delay
                    scheduleNextJob(emptyQueuePollInterval)
                }
            )
        }
    }

    private fun scheduleNextJob(delayMs: Long) {
        scheduledTask?.cancel(false)
        scheduledTask = executor.schedule({
            processNextJob()
        }, delayMs, TimeUnit.MILLISECONDS)
    }

    private fun processNextJob() {
        if (!isRunning.get()) return
        if (isPaused.get()) return

        // Check batch limit
        if (maxJobsPerBatch > 0 && processedInBatch >= maxJobsPerBatch) {
            processedInBatch = 0
            scheduleNextJob(emptyQueuePollInterval)
            return
        }

        // Process one job
        makeRequest("/_native/queue/work", "POST") { result ->
            if (!isRunning.get()) return@makeRequest

            result.fold(
                onSuccess = { json -> handleWorkResponse(json) },
                onFailure = {
                    scheduleNextJob(emptyQueuePollInterval)
                }
            )
        }
    }

    private fun handleWorkResponse(json: JSONObject) {
        val processed = json.optBoolean("processed", false)
        val pending = json.optInt("pending", 0)

        pendingJobCount.set(pending)

        if (processed) {
            processedInBatch++
            totalProcessed++

            // Extract job info
            val jobData = json.optJSONObject("job")
            val jobId = jobData?.optString("id", "unknown") ?: "unknown"
            val jobName = jobData?.optString("name", "unknown") ?: "unknown"
            val queue = jobData?.optString("queue", "default") ?: "default"
            val durationMs = json.optInt("duration_ms", 0)

            // Check for error (job failed but was processed)
            val error = json.optJSONObject("error")
            if (error != null) {
                val errorMessage = error.optString("message", "Unknown error")

                // Dispatch failure event
                dispatchEvent("Native\\Mobile\\Queue\\Events\\JobFailed", JSONObject().apply {
                    put("jobId", jobId)
                    put("jobName", jobName)
                    put("error", errorMessage)
                    put("willRetry", true)
                    put("attempts", jobData?.optInt("attempts", 1) ?: 1)
                    put("pending", pending)
                })
            } else {
                // Dispatch success event
                dispatchEvent("Native\\Mobile\\Queue\\Events\\JobCompleted", JSONObject().apply {
                    put("jobId", jobId)
                    put("jobName", jobName)
                    put("queue", queue)
                    put("durationMs", durationMs)
                    put("pending", pending)
                })
            }

            // More jobs? Process next one
            if (pending > 0) {
                scheduleNextJob(minDelayBetweenJobs)
            } else {
                // Dispatch queue empty event
                dispatchEvent("Native\\Mobile\\Queue\\Events\\QueueEmpty", JSONObject().apply {
                    put("queue", queue)
                    put("processedCount", totalProcessed)
                })

                // Poll again later in case new jobs arrive
                scheduleNextJob(emptyQueuePollInterval)
            }
        } else {
            // Queue was empty
            scheduleNextJob(emptyQueuePollInterval)
        }
    }

    // MARK: - Helpers

    private fun makeRequest(
        path: String,
        method: String,
        callback: (Result<JSONObject>) -> Unit
    ) {
        val bridge = phpBridge
        if (bridge == null) {
            callback(Result.failure(IllegalStateException("PHPBridge not initialized")))
            return
        }

        executor.execute {
            try {
                val request = PHPRequest(
                    url = path,
                    method = method,
                    headers = mapOf(
                        "Accept" to "application/json",
                        "Content-Type" to "application/json"
                    )
                )

                val response = bridge.handleLaravelRequest(request)

                // Parse response - extract body from HTTP response
                val parts = response.split("\r\n\r\n", limit = 2)
                if (parts.size >= 2) {
                    val body = parts[1]
                    val json = JSONObject(body)
                    callback(Result.success(json))
                } else {
                    callback(Result.failure(Exception("Invalid response format")))
                }
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }

    private fun dispatchEvent(event: String, payload: JSONObject) {
        val activity = activityProvider?.invoke() ?: return
        mainHandler.post {
            NativeActionCoordinator.dispatchEvent(activity, event, payload.toString())
        }
    }
}
